nelsk-kore/lib/kore/comm/platform_adapter.rb

```diff
@@ -0,0 +1,74 @@
+module Kore
+  module Comm
+    class PlatformAdapter < Kore::Machinery::Base
+      attr_accessor :engine
+      TRIGGER_PREFIX = '!'
+
+      def message_received(identity, raw)
+        self.log.debug "PlatformAdapter#message_received"
+        self.log.debug "  identity: #{identity}"
+        self.log.debug "  raw: #{raw}"
+
+        if self.is_cmd(raw)
+          self.log.debug "Command matched, parsing..."
+          begin
+            m = self.parse(raw)
+            m.originator = {
+              identity: identity,
+              platform: self.name
+            }
+          rescue Exception => e
+            self.log.warn "Error occurred while parsing message"
+            self.log.warn raw
+            self.log.warn e.message
+          else
+            @engine.route_ingress(m)
+          end
+        end
+      end
+
+      #############################################################
+      # Interface methods to be overriden
+      #############################################################
+      def send_message(emsg)
+        raise "platform adapter #{self.name} must implement \"send_message\" method"
+      end
+
+      def name
+        raise 'platform adapters must implement "name" method'
+      end
+
+      # Must be non-blocking!
+      def listen
+        raise "platform adapter #{self.name} must implement \"listen\" method"
+      end
+      #############################################################
+
+      def is_cmd(raw)
```

Comment:
Is this the platform adapter's responsibility? I feel like that should be something comm does.

Unless, of course, I'm misunderstanding this code, which I suspect I might be.

Reply:
I think so. Commands may look different based on the raw payload received from a service client, so it's up to the platform adapter to determine if that's actually a command it needs to pay attention to. If anything, it might need to end up an abstract method the concrete adapters define.